### PR TITLE
Consolidate and improve hello examples for numpy

### PR DESCRIPTION
**Pull request by nvkevlu**
Consolidate and improve hello examples for numpy and add new Recipe API.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/hello-world/hello-fedavg-numpy/recipe-api-approach/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+numpy>=1.21.0
+nvflare>=2.5.0
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/recipe-api-approach/requirements.txt]: or just "nvflare>=2.7.0rc

File: examples/hello-world/hello-fedavg-numpy/README.md [hunk 2]
```diff
@@ -1,36 +1,95 @@
...
+    └── hello_numpy_sag.ipynb
 ```
 
-### 3. Access the logs and results
+## Choose Your Approach
+
+### Recipe API (Recommended)
+- **Best for**: New users, modern development
+- **Files**: `client.py`, `model.py`, `job.py`
+- **Why**: Clean, simple, follows FLARE best practices
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/README.md]: Clean and simple to use,

I think other approaches are also following our best practices

File: examples/hello-world/hello-fedavg-numpy/README.md [hunk 3]
```diff
@@ -1,36 +1,95 @@
...
+
+### Job API (Learning)
+- **Best for**: Understanding FLARE internals
+- **Files**: Scripts + Jupyter notebooks
+- **Why**: Interactive tutorials, detailed explanations
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/README.md]: we could have a ipynb for the Recipe API as well, so I think this why is not valid.

should be something like flexible Python way to configure your job or something?

File: examples/hello-world/hello-fedavg-numpy/README.md [hunk 4]
```diff
@@ -1,36 +1,95 @@
...
 
-You can find the running logs and results inside the simulator's workspace:
+### Job Config (Reference)
+- **Best for**: Understanding job configuration, legacy migration
+- **Files**: JSON configuration files + job structure
+- **Why**: Shows traditional FLARE job configuration patterns with `config_fed_client.json` and `config_fed_server.json`
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/README.md]: is this Why means why this example is here?
or why use job config approach?

File: examples/hello-world/hello-fedavg-numpy/README.md [hunk 5]
```diff
@@ -1,36 +1,95 @@
...
 ```
 
-For how to use the FLARE API to run this app, see [this notebook](hello-fedavg-numpy_flare_api.ipynb).
+## What Each Approach Provides
+
+### Recipe API Approach
+- **Modern Python code**: Clean `client.py`, `model.py`, `job.py` structure
+- **Custom NumPy recipe**: `NumpyFedAvgRecipe` designed specifically for NumPy models
+- **Easy to understand**: Follows the same pattern as other hello-world examples
+- **Best for**: New users, modern development, learning FLARE concepts
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/README.md]: is this the best to learn FLARE concept, not sure, as this is easy and simple but hide some stuff behind

File: examples/hello-world/hello-fedavg-numpy/README.md [hunk 6]
```diff
@@ -1,36 +1,95 @@
...
+### Job API Approach  
+- **Interactive learning**: Jupyter notebooks with step-by-step tutorials
+- **Detailed explanations**: Shows how to build jobs programmatically
+- **FLARE internals**: Understanding of `FedJob`, `ScriptRunner`, and client API
+- **Best for**: Learning FLARE internals, understanding job configuration
+
+### Job Config Approach
+- **Traditional structure**: Complete FLARE job with JSON configuration files
+- **Configuration examples**: `config_fed_client.json` and `config_fed_server.json`
+- **Legacy patterns**: Shows how FLARE jobs were traditionally configured
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/README.md]: you can still configure it this way, and our job api or recipe api in the end generate the configs, in that regard it is the present way not legacy?

File: examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+## What This Example Does
+
+This is a simple federated learning example using NumPy that demonstrates how multiple clients can collaboratively train a model without sharing their data. The model starts with weights `[[1, 2, 3], [4, 5, 6], [7, 8, 9]]` and each client adds 1 to each weight during training, showing how federated averaging works.
+
+## How It Works
+
+### NVIDIA FLARE Job API
+
+The `fedavg_script_runner_hello-numpy.py` script builds the job with the Job API. Key components:
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md]: if this script same as recipe-api-approache/client.py maybe we use the same name "client.py"?

File: examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+Open Jupyter Lab for detailed tutorials:
+
+```bash
+jupyter lab
+```
+
+Then open:
+- `hello-fedavg-numpy_getting_started.ipynb` - Step-by-step tutorial
+- `hello-fedavg-numpy_flare_api.ipynb` - FLARE API examples
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md]: not sure if the flare_api notebook should be inside here, that one could also work with recipe api or even the config (after the job is submitted)

File: examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+## Interactive Learning
+
+Open Jupyter Lab for detailed tutorials:
+
+```bash
+jupyter lab
+```
+
+Then open:
+- `hello-fedavg-numpy_getting_started.ipynb` - Step-by-step tutorial
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/job-api-approach/README.md]: actually, should we put notebook here?
if so we might want to do same thing for recipe api?
or we should both do python code?

File: examples/hello-world/hello-fedavg-numpy/job-config-approach/README.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+- **`hello_numpy_sag.ipynb`**: Interactive notebook example
+
+## Installation
+
+Follow the [Installation](../../getting_started/README.md) instructions:
+
+```bash
+pip install nvflare
+pip install -r requirements.txt
+```
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/job-config-approach/README.md]: it is kind of weird we have quickstart doing install and run, then we explain install here again

File: examples/hello-world/hello-fedavg-numpy/job-config-approach/README.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+> **_NOTE:_** This example uses a Numpy-based trainer and will generate its data within the code.
+
+## Quick Start
+
+```bash
+pip install -r requirements.txt
+nvflare simulator -w /tmp/nvflare/hello-numpy-sag -n 2 -t 2 jobs/hello-numpy-sag
+```
+
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/job-config-approach/README.md]: maybe move the running part after installation

File: examples/hello-world/hello-fedavg-numpy/recipe-api-approach/model.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        Initialize the model with random weights.
+
+        Args:
+            input_size: Size of input layer
+            hidden_size: Size of hidden layer
+            output_size: Size of output layer
+        """
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.output_size = output_size
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/recipe-api-approach/model.py]: we are not using these arguments

File: examples/hello-world/hello-fedavg-numpy/recipe-api-approach/model.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    def evaluate(self):
+        """
+        Evaluate the model and return metrics.
+
+        Returns:
+            Dictionary containing evaluation metrics
+        """
+        # Simple evaluation: return the mean of the weights as accuracy
+        accuracy = np.mean(self.weights)
+        return {"accuracy": accuracy}
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/recipe-api-approach/model.py]: this is a bit weird though, to use mean as accuracy

File: examples/hello-world/hello-fedavg-numpy/recipe-api-approach/numpy_fedavg_recipe.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+from typing import Any, Optional
+
+from pydantic import BaseModel
+
+from nvflare.apis.dxo import DataKind
+from nvflare.app_common.abstract.aggregator import Aggregator
+from nvflare.app_common.aggregators import InTimeAccumulateWeightedAggregator
+from nvflare.app_common.shareablegenerators import FullModelShareableGenerator
+from nvflare.app_common.workflows.scatter_and_gather import ScatterAndGather
+from nvflare.app_opt.pt.job_config.base_fed_job import BaseFedJob
```
Comment (YuanTingHsieh) [inline examples/hello-world/hello-fedavg-numpy/recipe-api-approach/numpy_fedavg_recipe.py]: let's not import a job from PT, since that's optional dependencies

**Review by YuanTingHsieh — DISMISSED** [review]

File: examples/hello-world/hello-fedavg-numpy/recipe-api-approach/numpy_fedavg_recipe.py [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+from nvflare.client.config import ExchangeFormat, TransferType
+from nvflare.job_config.script_runner import FrameworkType, ScriptRunner
```
Comment (chesterxgchen) [inline examples/hello-world/hello-fedavg-numpy/recipe-api-approach/numpy_fedavg_recipe.py]: 1) The recipe if we decided to keep it here. it should be called server.py 
2) we probably should put this recipe in the nvflare.app_common.np.recipes folder

**Review by chesterxgchen — DISMISSED** [review]
I don't think we are list all three different approaches. 
It will ask customer to learn too much. Just one recipe approach. 
No need to emphasize fedavg in folder,  
hello-fedavg-numpy ==> hello-numpy

**Comment by chesterxgchen** [comment]
1) make sure you format the code and optimized import
2) make sure your RST file matching the code changes

File: examples/hello-world/hello-numpy/server.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    parser.add_argument("--num_rounds", type=int, default=3)
+    parser.add_argument("--learning_rate", type=float, default=1.0)
+
+    return parser.parse_args()
+
+
+def main():
+    args = define_parser()
+
+    n_clients = args.n_clients
```
Comment (chesterxgchen) [inline examples/hello-world/hello-numpy/server.py]: This file should be called job.py

File: docs/examples/hello_numpy.rst [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+You can follow :ref:`getting_started` on the general concept of setting up a
+Python virtual environment (the recommended environment) and how to install NVIDIA FLARE.
+
+Introduction
+-------------
+
+This tutorial is meant solely to demonstrate how the NVIDIA FLARE system works, without introducing any actual deep
+learning concepts.
+
+Through this exercise, you will learn how to use NVIDIA FLARE with NumPy to perform basic
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: prefer to organize this follow the structure similar to hello-pt, including 
# code structure 
# data 
# model 
# client side
# server side
# Job Recipe 
# Run Job

and the rest of docs

File: docs/examples/hello_numpy.rst [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+
+Model
+-----
+
+The ``SimpleNumpyModel`` class implements a basic neural network model using NumPy arrays:
+
+.. literalinclude:: ../../examples/hello-world/hello-numpy/model.py
+   :language: python
+   :linenos:
+   :caption: model.py
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: :lines: 14-

Add above line to remove copy right header

File: docs/examples/hello_numpy.rst [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+Client Side
+-----------
+
+The training script ``client.py`` is the main script that will be run on the clients. It contains the NumPy specific
+logic for training:
+
+.. literalinclude:: ../../examples/hello-world/hello-numpy/client.py
+   :language: python
+   :linenos:
+   :caption: client.py
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: :lines: 14-

File: docs/examples/hello_numpy.rst [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+Job Recipe
+----------
+
+The ``job.py`` script uses the modern Job Recipe API to define the federated learning job:
+
+.. literalinclude:: ../../examples/hello-world/hello-numpy/job.py
+   :language: python
+   :linenos:
+   :caption: job.py
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: :lines: 14-

File: docs/examples/hello_numpy.rst [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+To run this example:
+
+1. Clone the repository and navigate to the example directory:
+
+.. code-block:: shell
+
+   $ git clone https://github.com/NVIDIA/NVFlare.git
+   $ cd NVFlare/examples/hello-world/hello-numpy
+
+2. Install the required dependencies:
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: this should go to the top again see hello-pt.rst

**Review by chesterxgchen — COMMENTED** [review]
few minor changes

File: docs/examples/hello_numpy.rst [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+Use the following command in your terminal to start the job with the specified number of rounds and number of clients.
+
+.. code-block:: text
+
+   python job.py --num_rounds 3 --n_clients 2
+
+The full source code for this exercise can be found in
+:github_nvflare_link:`examples/hello-world/hello-numpy <examples/hello-world/hello-numpy/>`.
+
+Previous Versions of Hello NumPy
```
Comment (chesterxgchen) [inline docs/examples/hello_numpy.rst]: L127-135  we don't need this section

**Review by chesterxgchen — APPROVED** [review]
